### PR TITLE
Bk/add vertical layout direction target content offset

### PR DESCRIPTION
**Pull request by bryankeller**
## Details

This PR refactors how `MagazineLayout` attempts to maintain the scroll position when batch updates occur and when the bounds changes (on device rotate, for example). A new configuration parameter has been added to control the vertical layout direction.

```swift
/// The vertical layout direction of items in the collection view. This property changes the behavior of scroll-position-preservation when
/// performing batch updates or when the collection view's bounds changes.
public enum MagazineLayoutVerticalLayoutDirection {

  /// The vertical layout direction of items goes from top to bottom. Items inserted at the top push content down. When scrolled to the
  /// top, inserted items at the top will be visible without scrolling. On bounds change (like device rotation or window size change), the
  /// topmost visible item will remain visible while items at the bottom may change their position in relation to the bottom edge.
  ///
  /// Using this layout direction is essentially the default layout direction of all collection views.
  case topToBottom

  /// The vertical layout direction of items goes from bottom to top. Items inserted at the bottom push content up. When scrolled to the
  /// bottom, inserted items at the bottom will be visible without scrolling. On bounds change (like device rotation or window size
  /// change), the bottommost visible item will remain visible while items at the top may change their position in relation to the top
  /// edge.
  ///
  ///  You might consider using this layout direction when building a message thread UI.
  case bottomToTop
}
```

#### Top to bottom behavior

https://github.com/airbnb/MagazineLayout/assets/746571/cfec97e4-d4d8-4cb7-a185-2649a20f5df2

#### Bottom to top behavior

https://github.com/airbnb/MagazineLayout/assets/746571/8bdaf680-37e6-4d88-9b9c-b6057f4faea5

## Related Issue

N/A

## Motivation and Context

Support better scroll position behavior for message thread UIs

## How Has This Been Tested

Unit tests, example app

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: MagazineLayout/Public/Types/MagazineLayoutVerticalLayoutDirection.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// The vertical layout direction of items in the collection view. This property changes the behavior of scroll-position-preservation when
+/// performing batch updates or when the collection view's bounds changes.
+public enum MagazineLayoutVerticalLayoutDirection {
+
+  /// The vertical layout direction of items goes from top to bottom. Items inserted at the top push content down. When scrolled to the
```
Comment (brynbodayle) [inline MagazineLayout/Public/Types/MagazineLayoutVerticalLayoutDirection.swift:20]: mention that this is the standard way UICollectionViewLayouts work?

File: MagazineLayout/LayoutCore/Types/TargetContentOffsetAnchor.swift [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+
+// MARK: - TargetContentOffsetAnchor
+
+enum TargetContentOffsetAnchor: Equatable {
```
Comment (brynbodayle) [inline MagazineLayout/LayoutCore/Types/TargetContentOffsetAnchor.swift:20]: could we explain what this does a bit in some doc comments?

**Review by brynbodayle — APPROVED** [review]
Nice!

File: MagazineLayout/Public/Types/MagazineLayoutVerticalLayoutDirection.swift (hunk 1 above)
Comment (bryankeller) [reply MagazineLayout/Public/Types/MagazineLayoutVerticalLayoutDirection.swift:20]: I basically say that here https://github.com/airbnb/MagazineLayout/pull/114/files#diff-a65dd25675c8fc47b5b55e8f3d48d4f20b50d15929e0269692e506ec8aa7cc4bR24

**Pull request merged**